WooCommerce/Classes/Notifications/PushNotificationsManager.swift

```diff
@@ -220,46 +229,72 @@ extension PushNotificationsManager {
+            return UNNotificationPresentationOptions(rawValue: 0)
         }
 
-        // Badge: Reset
-        guard userInfo.string(forKey: APNSKey.type) != PushType.badgeReset else {
-            return
+        handleRemoteNotificationInAllAppStates(content.userInfo)
+
+        if let foregroundNotification = PushNotification.from(userInfo: content.userInfo) {
+            configuration.application
+                .presentInAppNotification(title: foregroundNotification.title,
+                                          subtitle: foregroundNotification.subtitle,
+                                          message: foregroundNotification.message,
+                                          actionTitle: Localization.viewInAppNotification) { [weak self] in
+                    guard let self = self else { return }
+                    self.presentDetails(for: foregroundNotification)
+                    self.foregroundNotificationsToViewSubject.send(foregroundNotification)
+                    ServiceLocator.analytics.track(.viewInAppPushNotificationPressed,
+                                                   withProperties: [AnalyticKey.type: foregroundNotification.kind.rawValue])
+                }
+
+            foregroundNotificationsSubject.send(foregroundNotification)
         }
 
-        // Analytics
-        trackNotification(with: userInfo)
+        _ = await synchronizeNotifications()
+        return UNNotificationPresentationOptions(rawValue: 0)
+    }
 
-        // Handling!
-        let handlers = [
-            handleSupportNotification,
-            handleForegroundNotification,
-            handleInactiveNotification,
-            handleBackgroundNotification
-        ]
-
-        for handler in handlers {
-            if handler(userInfo, completionHandler) {
-                break
-            }
+    @MainActor
+    func handleUserResponseToNotification(response: UNNotificationResponse) async {
+        // Remote notification response is handled separately.
+        if let notification = PushNotification.from(userInfo: response.notification.request.content.userInfo) {
+            handleRemoteNotificationInAllAppStates(response.notification.request.content.userInfo)
+            return await handleInactiveRemoteNotification(notification: notification)
+        } else {
+            localNotificationResponsesSubject.send(response)
+        }
+    }
+
+    /// Handles a Notification while in Background Mode
+    ///
+    /// - Parameters:
+    ///     - userInfo: The Notification's Payload
+    ///     - completionHandler: A callback, to be executed on completion
+    ///
+    /// - Returns: True when handled. False otherwise
+    ///
+    @MainActor
+    func handleRemoteNotificationInTheBackground(userInfo: [AnyHashable: Any]) async -> UIBackgroundFetchResult {
+        guard applicationState == .background, let _ = userInfo[APNSKey.identifier] else {
```

Comment:
I think it will help to add comments about each condition check. Also, a comment about why returning `.noData`. 
I am thinking about something like the following. (I am sure that we can make it better. 😅 )
```swift
guard  applicationState == .background, // Proceeding only if the app is in background. 
            let _ = userInfo[APNSKey.identifier] // making sure that we are only processing a push notification
 ````
 
 I am new to `UIBackgroundFetchResult` and please ignore my comment if this will be obvious for someone who has prior experience.